freestyle/shared/src/main/scala/freestyle/module.scala

```diff
@@ -28,94 +24,65 @@ object materialize {
-          val dependentTypeCtor = m.returnType.typeConstructor
-          extractCoproductSymbol(companionClass, dependentTypeCtor)
-        case _ =>
-          Nil
-      }
-    }
+      def isFreeStyleModule(cs: ClassSymbol): Boolean =
+        cs.baseClasses.exists {
+          case x: ClassSymbol => x.name == TypeName("Modular")
+        }
 
-    def findAlgebras(s: Type): List[TermName] = {
-      def loop(l: List[Symbol], acc: List[TermName]): List[TermName] = {
-        l match {
-          case Nil => acc
-          case h :: t =>
-            val companionClass    = h.asMethod.returnType.companion.typeSymbol.asClass
-            val dependentTypeCtor = h.asMethod.returnType.typeConstructor
-            val coproducts        = extractCoproductSymbol(companionClass, dependentTypeCtor)
-            loop(t, acc ++ coproducts)
+      def fromClass(c: ClassSymbol) : List[Type] =
+        c.companion.info.decls.toList.flatMap {
+          case x: MethodSymbol => fromMethod(x.asMethod)
+          case _ => Nil
         }
+
+      def fromMethod(meth: MethodSymbol): List[Type] = {
+        val companionClass    = meth.returnType.companion.typeSymbol.asClass
+        if (isFreeStyleModule(companionClass))
+          fromClass(companionClass)
+        else
+          List( meth.returnType.typeConstructor)
+      }
+
+      s.companion.info.decls.toList.flatMap {
+        case x: MethodSymbol if x.isAbstract => fromMethod(x.asMethod)
+        case _ => Nil
       }
-      loop(s.decls.toList.filter(_.isAbstract), Nil)
     }
 
-    def cp(n: TermName) =
-      Select(Ident(n), TypeName("Op"))
-
-    def mkModuleCoproduct(algebras: List[TermName]): List[String] = { //ugly hack because as String it does not typecheck early which we need for types to be in scope
-      val result = algebras match {
-        case List(el) =>
-          (0, q"type Op[A] = cats.data.Coproduct[$el.Op, Nothing, A]") :: Nil //TODO this won't work but we need a solution to single service modules
-        case l @ _ :: _ =>
-          l.scanLeft[(Int, AnyRef), List[(Int, AnyRef)]]((0, q"")) {
-            case ((pos, acc), el) =>
-              val cpName = if (pos + 1 != l.size) TypeName(s"C$pos") else TypeName("Op")
-              val newTree = (acc, el) match {
-                case (q"", r: TermName) => r
-                case (l: TermName, r: TermName) =>
-                  q"type $cpName[A] = cats.data.Coproduct[${cp(l)}, ${cp(r)}, A]"
-                case (l: TypeDef, r: TermName) =>
-                  q"type $cpName[A] = cats.data.Coproduct[${cp(r)}, ${l.name}, A]"
-                case x =>
-                  fail(
-                    s"found unexpected case building Coproduct: $x with types ${x.map(_.getClass)}")
-              }
-              (pos + 1, newTree)
-          }
 
+    def mkCoproduct(algebras: List[Type]): List[TypeDef] =
+      algebras.map(x => TermName(x.toString) ) match {
+        case Nil => List( q"type Op[A] = Nothing")
```

Comment:
As far as possible (I may have made mistakes), I have tried to stick to the previous operation. This is what the code before the PR was doing for the case with no algebras. It can be changed, of course.